src/common/get-card-selection-view-data.ts

```diff
@@ -1,65 +1,147 @@
     storeData: CardSelectionStoreData,
+    unifiedScanResultStoreData: UnifiedScanResultStoreData,
+    getUnavailableHighlightStatus: GetUnavailableHighlightStatus,
+) => CardSelectionViewData;
+
+export const getCardSelectionViewData: GetCardSelectionViewData = (
+    cardSelectionStoreData: CardSelectionStoreData,
+    unifiedScanResultStoreData: UnifiedScanResultStoreData,
+    getUnavailableHighlightStatus: GetUnavailableHighlightStatus,
 ): CardSelectionViewData => {
     const viewData = getEmptyViewData();
 
-    if (!storeData) {
+    if (
+        isEmpty(cardSelectionStoreData) ||
+        isEmpty(unifiedScanResultStoreData) ||
+        unifiedScanResultStoreData.results == null
+    ) {
         return viewData;
     }
 
-    viewData.visualHelperEnabled = storeData.visualHelperEnabled || false;
+    viewData.visualHelperEnabled = cardSelectionStoreData.visualHelperEnabled || false;
+    viewData.expandedRuleIds = getRuleIdsOfExpandedRules(cardSelectionStoreData.rules);
 
-    viewData.expandedRuleIds = getRuleIdsOfExpandedRules(storeData.rules);
+    const allResultUids = getAllResultUids(cardSelectionStoreData.rules);
+    const resultsWithUnavailableHighlightStatus = getResultsWithUnavailableHighlightStatus(
+        unifiedScanResultStoreData,
+        getUnavailableHighlightStatus,
+    );
+
+    allResultUids.forEach(resultUid => {
+        viewData.resultsHighlightStatus[resultUid] = 'hidden';
+    });
 
-    if (!storeData.visualHelperEnabled) {
-        // no selected cards; no highlighted instances
+    if (!cardSelectionStoreData.visualHelperEnabled) {
+        // no selected cards; no highlighted instances and unavailable instances shown as such.
+        viewData.resultsHighlightStatus = {
+            ...viewData.resultsHighlightStatus,
+            ...resultsWithUnavailableHighlightStatus,
+        };
         return viewData;
     }
 
+    // if no rules are expanded, highlight all results except for those unavailable.
     if (viewData.expandedRuleIds.length === 0) {
-        viewData.highlightedResultUids = getAllResultUids(storeData.rules);
+        forOwn(viewData.resultsHighlightStatus, (_, resultUid) => {
+            viewData.resultsHighlightStatus[resultUid] = 'visible';
+        });
+        viewData.resultsHighlightStatus = {
+            ...viewData.resultsHighlightStatus,
+            ...resultsWithUnavailableHighlightStatus,
+        };
         return viewData;
     }
 
+    // if there are any cards selected, make sure they are highlighted if not unavailable.
     viewData.selectedResultUids = getOnlyResultUidsFromSelectedCards(
-        storeData.rules,
+        cardSelectionStoreData.rules,
         viewData.expandedRuleIds,
     );
 
-    viewData.highlightedResultUids = viewData.selectedResultUids.length
-        ? viewData.selectedResultUids
-        : getAllResultUidsFromRuleIdArray(storeData.rules, viewData.expandedRuleIds);
+    if (viewData.selectedResultUids.length) {
```

Comment:
I think it would be a little more clear to say `if (viewData.selectedResultUids.length > 0) {` (I never trust my rememberance of javascript's falsiness rules and had to test it to convince myself this was right)

Reply:
Agreed. I actually did have that at some point; I'll re-add it! 😄